cmake: 3.31.7 -> 4.1.1

biodranik opened the pull request:
Updated to the latest official cmake release: https://cmake.org/cmake/help/latest/release/4.1.html

A side note: release notes do not mention [UNITY_BUILD_RELOCATABLE](https://cmake.org/cmake/help/latest/prop_tgt/UNITY_BUILD_RELOCATABLE.html) target property introduced in 4.0.
It fixes the remote ccache hit rate when unity builds are enabled. Both unity builds and ccache are greatly speeding up C/C++/ObjC/Cuda builds when enabled.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

biodranik committed:
cmake: Remove obsolete patch

Cygwin patch does not apply anymore, it was done for a very old cmake 3.2.2

MisileLab commented:
1. Base branch should be staging because it involves many rebuilds.
2. https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#writing-good-commit-messages

biodranik committed:
cmake: 3.31.7 -> 4.1.1

- Update cmake to the latest 4.1.1 release:
  https://cmake.org/cmake/help/latest/release/4.1.html
- FindCurl patch has already been merged:
  https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9900
- Darwin `-rpath` patch is replaced by the binary paths patch
- Fix previous patches for a newer source code

biodranik commented:
@MisileLab thanks! Rebased onto staging, split commits, updated patches. PTAL

biodranik commented:
Build failure is not relevant, what to do in such cases?
```
Error: No space left on device : '/home/runner/runners/2.323.0/_diag/blocks/0cdc5f65-806c-4a6a-a4b2-194e18d5d891_2db42233-a5e5-50d5-4ab0-e31ade3450a6.9'
```

biodranik committed:
xcbuild: Fix build errors after cmake 4 upgrade

The fix is required for cmake 4+ to fix this build error:
> Compatibility with CMake < 3.5 has been removed from CMake.

As xcbuild repo is archived, there is no any other way to fix it upstream.

emilazy commented:
Since this is a mass breaking change we should most likely do it after 25.05’s release rather than right before the freeze.

biodranik commented:
@emilazy only old packages that explicitly declare cmake_minimum_required < 3.5 are failing (2 of them so far), so it should be easy to workaround. Cacheable unity builds are important to speed up development, it would be also great to integrate it into nix packages to speed up constant rebuilds.

MisileLab commented:
cmake 4.0.1 released

biodranik commented on pkgs/by-name/cm/cmake/package.nix:
```diff
@@ -50,11 +50,11 @@ stdenv.mkDerivation (finalAttrs: {
   src = fetchurl {
     url = "https://cmake.org/files/v${lib.versions.majorMinor finalAttrs.version}/cmake-${finalAttrs.version}.tar.gz";
-    hash = "sha256-ZTQn8PUBR1Cq//InJ/sqpgxscyypGAjPt4ziLd2eVfA=";
+    hash = lib.fakeHash;
```
1. What are the ways to calculate the updated hash here?
2. What is the right command to check this change locally?

LordGrimmauld replied:
typically: put an empty (or fake) hash, see how it crashes, replace.
You might be able to do some tricks with `nix-prefetch-url https://cmake.org/files/v4.0/cmake-4.0.0.tar.gz`, but iirc that only returns the old hash format, and i don't know how to convert.

biodranik replied:
Thanks! How to "see how it crashes"? Which command should be used to test if the change works locally?

Also, any ideas why it fails [like this](https://github.com/NixOS/nixpkgs/actions/runs/14546787010/job/40813228286#step:4:791)? Is it because of the fakeHash?
```
… while evaluating attribute 'src' of derivation 'cmake-minimal-4.0.1'
         at /nix/store/m5rpfz1c0iqjsn3r9d4j1g96g0xr5viq-source/pkgs/by-name/cm/cmake/package.nix:55:3:
           54|
           55|   src = fetchurl {
             |   ^
           56|     url = "[https://cmake.org/files/v${lib.versions.majorMinor](https://cmake.org/files/v$%7Blib.versions.majorMinor) finalAttrs.version}/cmake-${finalAttrs.version}.tar.gz";

       … while calling the 'getAttr' builtin
         at <nix/derivation-internal.nix>:50:17:
           49|     value = commonAttrs // {
           50|       outPath = builtins.getAttr outputName strict;
             |                 ^
           51|       drvPath = strict.drvPath;

       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:37:12:
           36|
           37|   strict = derivationStrict drvAttrs;
             |            ^
           38|

       … while evaluating derivation 'cmake-4.0.1.tar.gz'
         whose name attribute is located at /nix/store/m5rpfz1c0iqjsn3r9d4j1g96g0xr5viq-source/pkgs/stdenv/generic/make-derivation.nix:439:13

       … while evaluating attribute 'SSL_CERT_FILE' of derivation 'cmake-4.0.1.tar.gz'
         at /nix/store/m5rpfz1c0iqjsn3r9d4j1g96g0xr5viq-source/pkgs/build-support/fetchurl/default.nix:232:5:
          231|     # needed to access the resource
          232|     SSL_CERT_FILE =
             |     ^
          233|       if

       … while evaluating a path segment
         at /nix/store/m5rpfz1c0iqjsn3r9d4j1g96g0xr5viq-source/pkgs/build-support/fetchurl/default.nix:242:10:
          241|       then
          242|         "${cacert}/etc/ssl/certs/ca-bundle.crt"
             |          ^
          243|       else

       error: cannot coerce null to a string: null
```

MisileLab replied:
> typically: put an empty (or fake) hash, see how it crashes, replace. You might be able to do some tricks with `nix-prefetch-url https://cmake.org/files/v4.0/cmake-4.0.0.tar.gz`, but iirc that only returns the old hash format, and i don't know how to convert.

do nix-prefetch-url, then `nix-hash --type sha256 --to-sri $hash`

MisileLab replied:
for this version, hash is `sha256-3cVK1juH4VPPUL5FCmWA8bF7SIHeiUHalj/1aZGkCDs=`

biodranik replied:
Thanks! The hash for 4.0.1 is although different on my machine:
```
$ nix-prefetch-url https://cmake.org/files/v4.0/cmake-4.0.1.tar.gz
path is '/nix/store/3kvkxl3jvg0fgk55bhxpl5sxmqkabg8a-cmake-4.0.1.tar.gz'
1q9119wg68vz3ki4g3yw3dkb90zpbr13vy2raar21rb31vhafc6n
$ nix-hash --type sha256 --to-sri 1q9119wg68vz3ki4g3yw3dkb90zpbr13vy2raar21rb31vhafc6n
sha256-1jCn4A5j5SCyUln4PUJe94O0Zhvcj0fiHH8j83gKIeE=
```

MisileLab replied:
> Thanks! The hash for 4.0.1 is although different on my machine:
> ```
> $ nix-prefetch-url https://cmake.org/files/v4.0/cmake-4.0.1.tar.gz
> path is '/nix/store/3kvkxl3jvg0fgk55bhxpl5sxmqkabg8a-cmake-4.0.1.tar.gz'
> 1q9119wg68vz3ki4g3yw3dkb90zpbr13vy2raar21rb31vhafc6n
> $ nix-hash --type sha256 --to-sri 1q9119wg68vz3ki4g3yw3dkb90zpbr13vy2raar21rb31vhafc6n
> sha256-1jCn4A5j5SCyUln4PUJe94O0Zhvcj0fiHH8j83gKIeE=
> ```

I accidentally posted 4.0.0's hash, your hash is right.

biodranik replied:
@MisileLab Ah, you wrote the hash for 4.0.0, not for 4.0.1. But why is it different from the one I used before for 4.0.0 `hash = "sha256-ZTQn8PUBR1Cq//InJ/sqpgxscyypGAjPt4ziLd2eVfA=";` which passed previous CI checks?

MisileLab replied:
🤔 I don't know, maybe I inputted the wrong url

biodranik commented:
One [check shows](https://logs.ofborg.org/?key=nixos/nixpkgs.394444&attempt_id=b13eae03-7612-4e00-99e5-31cadc84e0e3) "neutral" and failed due to a timeout for llvm compilation on darwin x86_64. What is usually done in such cases?
```
[2068/3749] Building CXX object lib/Target/Mips/MCTargetDesc/CMakeFiles/LLVMMipsDesc.dir/MipsMCCodeEmitter.cpp.o
error: building of '/nix/store/2a1lyv8bnp5x64sra8iwrvqmnlr0z97p-llvm-19.1.7.drv^dev' from .drv file timed out after 3600 seconds
```

biodranik commented:
Checks finished. The second "neutral" check also [failed in llvm](https://logs.ofborg.org/?key=nixos/nixpkgs.394444&attempt_id=2abca931-5eeb-4932-8f82-92977c96a08a), any suggestions about the next steps? Can this MR be merged now? 
```
[4460/4460] Linking CXX executable unittests/tools/llvm-mca/LLVMMCATests
buildPhase completed in 56 minutes 56 seconds
Running phase: checkPhase
check flags: -j2 check-all
[1/2] Preparing lit tests
[1/2] Running all regression tests
llvm-lit: /private/tmp/nix-build-llvm-19.1.7.drv-0/llvm-src-19.1.7/llvm/utils/lit/lit/llvm/config.py:508: note: using split-file: /private/tmp/nix-build-llvm-19.1.7.drv-0/llvm-src-19.1.7/llvm/build/bin/split-file
llvm-lit: /private/tmp/nix-build-llvm-19.1.7.drv-0/llvm-src-19.1.7/llvm/utils/lit/lit/llvm/config.py:508: note: using yaml2obj: /private/tmp/nix-build-llvm-19.1.7.drv-0/llvm-src-19.1.7/llvm/build/bin/yaml2obj
llvm-lit: /private/tmp/nix-build-llvm-19.1.7.drv-0/llvm-src-19.1.7/llvm/utils/lit/lit/llvm/config.py:508: note: using llvm-objcopy: /private/tmp/nix-build-llvm-19.1.7.drv-0/llvm-src-19.1.7/llvm/build/bin/llvm-objcopy
An error occurred retrieving the git revision: [Errno 2] No such file or directory: 'git'
-- Testing: 54377 tests, 2 workers --
error: building of '/nix/store/35mns62bgcxm5rbrrmgab2xz5v88rc3d-llvm-19.1.7.drv^out' from .drv file timed out after 3600 seconds
```

LordGrimmauld commented:
How long will cmake 3.x be supported? if the answer to that question is "longer than november 2025", i believe merging this so close to a release is a bad idea

emilazy commented:
Yeah, it didn’t seem like there was appetite to deal with compatibility breaks in the many projects setting old CMake versions before branch‐off. We’re already seemingly getting one fewer `staging-next` before release than hoped for. This should be able to land [somewhere around the 8th](https://github.com/nixos/nixpkgs/issues/390768) for 25.11.

emilazy commented:
(btw, thank you for Organic Maps! 💜)

biodranik commented:
Thanks for the feedback! Is it a normal practice to have a separate package for a major version upgrade, like cmake4 instead of cmake? If yes, what are the best examples to check and does it make sense to change this PR to a separate cmake4 package?

LordGrimmauld commented:
Side note: Doing a cmake update as *first contribution* is brave, you are doing well :)

emilazy commented:
> Assuming nothing major breaks,

CMake allows projects to set the maximum version of CMake they expect, and emulates the behaviour of old versions as required. 4.0 dropped compatibility with CMake < 3.5, so every project that sets an older version than that will error out. We can override the version, but in that case we’re subjecting projects to every breaking change between their declared version and 3.5, which could cause issues. CMake has been issuing a warning for such old versions for a while, and my experience working across Nixpkgs has been that that warning shows up often enough that it’s going to be a lot of work to get packages building again even if the breaking changes don’t actually break them (silently or loudly).

I do think that it makes sense to just update it in‐place and deal with the fallout, though; better to move the ecosystem forward when we can. It’s just that introducing something like this days before the breaking change deadline will take up a lot of resources that are currently occupied preparing for the release, and risk delaying it. The [staging cycle](https://github.com/NixOS/nixpkgs/blob/6c18fdbbaca6d4b4ae5efb757b1cc75e1fcdf52f/CONTRIBUTING.md#staging) means that we handle a batch of world‐rebuilding and potentially breaking changes in one go, and that if we have insufficient resources to triage and fix things it can delay landing security fixes or making the release deadline. It also means that at some point we call it good enough and merge it in to hit the channels, at which point the effort of fixing remaining issues becomes more distributed, but as we’re about to head into the Zero Hydra Failures effort we generally hope to be decreasing the number of broken things. (Of course we’re always happy to have more people hop on Matrix and help out with staging cycles!)

It would, however, be acceptable to introduce a separate `cmake_4` in 25.05, if there is demand for it from packages or users, and then flip the default and drop the old version in 25.11. Generally we avoid that unless there’s particular reason, since it means additional complexity in package code to support multiple versions or dealing with copy‐paste, and it’s expected that stable releases don’t necessarily have the latest‐and‐greatest software when breaking changes are involved, but if there are sufficiently compelling features of CMake 4.0 that it’d hurt for stable users to not have it then it’s an option. Otherwise, I think it should be okay to land this in a couple weeks’ time; we like to land big breaking changes like compiler updates early in a new release cycle when we can.

BTW, we prefer the structured `(lib.cmakeFeature "CMAKE_POLICY_VERSION_MINIMUM" "3.5")` interface. Other than that the diff looks great to me, thanks!

biodranik commented:
Thanks! I've integrated the use of `lib.cmakeFeature` and updated toml11 to a fresher version, which already declares min cmake 3.5

> I do think that it makes sense to just update it in‐place and deal with the fallout, though; better to move the ecosystem forward when we can.

IMO it is the best approach. How can I reproduce/check which packages may fail due to incompatibility with cmake 4? I discovered only two of them and already fixed them in this PR.

LordGrimmauld commented:
> Thanks! I've integrated the use of `lib.cmakeFeature` and updated toml11 to a fresher version, which already declares min cmake 3.5
> 
> > I do think that it makes sense to just update it in‐place and deal with the fallout, though; better to move the ecosystem forward when we can.
> 
> IMO it is the best approach. How can I reproduce/check which packages may fail due to incompatibility with cmake 4? I discovered only two of them and already fixed them in this PR.

You'd want to eval all of nixpkgs, find all direct dependendents of cmake, extract their source archive, and grep for their minimum cmake version. Likely requires some custom tooling.

emilazy commented:
> IMO it is the best approach. How can I reproduce/check which packages may fail due to incompatibility with cmake 4? I discovered only two of them and already fixed them in this PR.

Realistically the easy ways to detect the big issues are:

1. Build entire system configurations with this change, or at least substantial packages with large dependency trees – which will take a rather long time, since throwing away `cmake` means you’ll be building a huge number of things from scratch.

2. Get this merged into `staging`, wait for `staging-next` to be cut, and then check [the reports](https://malob.github.io/nix-review-tools-reports/) to see packages break in real time, ordered by how many downstream users they block. Then fixes go to `staging-next`. (We only merge `staging-next` into `master` once the level of regressions looks acceptable, so users don’t immediately experience the issues the report flags up.)

We tend to do a mixture of both for large changes like this – some basic smoke testing for (1), though usually not as thorough as building entire configurations, and then switching to a reactive fixes mode for (2). (Joining the [Staging room](https://matrix.to/#/#staging:nixos.org) on Matrix helps coordinate the latter.)

That said, I think this is in an okay state to land as long as we can build `stdenv` on all platforms; we pretty much expect more things to break than can reasonably be addressed prior to merge, but we also expect that the effort of fixing each individual instance will generally be low, and I think you’ve done a sufficient amount of work on the actual version bump and smoke testing this. I’m not sure if the Linux `stdenv` uses CMake, but Darwin does because of LLVM; I can spin up a build to test the latter if needed.

biodranik commented:
Again [the same failure](https://logs.ofborg.org/?key=nixos/nixpkgs.394444&attempt_id=0c90460c-93e9-4449-b330-bf9a2f335fbe) of the long llvm build due to low timeout:
```
[2018/3749] Building CXX object lib/Target/ARM/MCTargetDesc/CMakeFiles/LLVMARMDesc.dir/ARMMCExpr.cpp.o
error: building of '/nix/store/f0fqf0zlj4x3r38w8gw83x2fl8y1j15n-llvm-19.1.7.drv^dev' from .drv file timed out after 3600 seconds
```

LordGrimmauld commented:
> Again [the same failure](https://logs.ofborg.org/?key=nixos/nixpkgs.394444&attempt_id=0c90460c-93e9-4449-b330-bf9a2f335fbe) of the long llvm build due to low timeout:

Timeouts happen. The timeout on ofborg is quite low due to computation power being limited.
Hydra apparently has a timeout of 12h. I wouldn't worry too much about the CI timing out here, if the package itself does actually build locally. Once we get past 25.05 branchoff, we'll validate the package (and some basic core dependencies to get a minimal system) do build.

Or, also proposed, run this against a dedicated hydra job set, though that would have to wait a little later as hydra compute is basically booked until 25.05 release.

MisileLab commented:
4.0.2 released: https://cmake.org/cmake/help/latest/release/4.0.html#id18

biodranik commented:
Thanks, updated to 4.0.2

MisileLab commented:
again, 4.0.3 released: https://cmake.org/cmake/help/latest/release/4.0.html#id3 (I guess no api change)

jys1670 commented:
Since CMake 4 is not really backward compatible with older versions and there are far too many breakages because of that, can we go with introducing a separate `cmake_4`?
Aside from speeding up builds, it finally has [support for `import std;` with GCC](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/10023).

jys1670 commented:
> How to discover these breakages and count them? Pushing upstream or nix maintainers to fix it looks like a better long-term strategy.

I guess it's about evaluating one of [these jobsets](https://hydra.nixos.org/project/nixpkgs) locally (with cmake changes applied) and comparing the number of failures to see how many more appeared. It's very resource-intensive... It might be possible to get some help with evaluation. I remember vcunat creating separate smaller jobsets for some PRs. It's worth asking around in nixos matrix.

> Does it work reliably? The MR references some issues, like [this one](https://gitlab.kitware.com/cmake/cmake/-/issues/26609).

I have tried compiling a test project (something like [this](https://www.reddit.com/r/cpp/comments/1j7vplc/gcc_support_std_module_with_cmake_40_now/)) on Arch, and it worked. Perhaps it was some packaging issue.

LordGrimmauld commented:
I wouldn't be against introducing a cmake_3 to pin stuff that can't be upgraded, but the default should be 4.x. Both to stay up to date and to actually find (and count) the unfixed stuff.
Finding breaks would basically be either running this PR in a hydra jobset, or yoloing it and fixing during staging cycle. Either way it'd involve hydra.

emilazy commented:
FWIW this is still on the radar and has been discussed recently in the Staging room on Matrix. Sorry for the radio silence here, it’s just a matter of people having limited time to devote to this stuff right now and Hydra being busy. This kind of update is one of the trickier ones to land without gumming up our processes entirely.

I think we’re agreed that the correct way forward is to set up a Hydra jobset with the bump and the long‐problematic `fixCmakeFiles` hook removed and work on triaging the regressions before sending this to `staging`. IIRC, the immediate blocker there is that CMake itself fails to build without `fixCmakeFiles` and nobody yet put the effort into figuring out why. If you felt like putting energy into dropping the hook and figuring out how to get CMake itself to build in this PR, it’d probably reduce the inertia of getting the jobset set up, but otherwise hopefully someone will have the time for it soon.  If you felt like joining us on Matrix to help with a triage effort then it would be a great help too, but you’ve already done a lot of work here so don’t feel obligated.

BTW, I think `-DCMAKE_POLICY_VERSION_MINIMUM=3.5` is still deprecated/produces warnings. Gentoo (among probably others) has been working on this transition and is setting 3.10, which I believe is the lowest that CMake haven’t indicated intent to deprecate. We should probably follow suit to avoid doing all of this again any time soon (though hopefully we won’t need `xcbuild` for too much longer now that Swift Build exists).

The `toml11` bump probably [still breaks the build of Nix itself](https://github.com/NixOS/nix/issues/11441) and pinging on that issue hasn’t produced results so far. But we can deal with that problem later; it shouldn’t be a blocker for the jobset.

MisileLab commented:
https://cmake.org/cmake/help/latest/release/4.1.html#id1 released

emilazy committed:
cmake: remove obsolete LLVM feature check workaround

This was fixed upstream in CMake 3.30.0 by
<https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9427>.

emilazy committed:
cmake: remove obsolete Large File Support flags

This was addressed upstream and the linked issue was closed.

emilazy committed:
cmake: remove obsolete `-Wno-elaborated-enum-base`

This was introduced for LLVM 11 in
982f7228ffd01450338384a81b31f6527d674574, but the package builds fine
with LLVM 21 without i.

emilazy committed:
zeromq: backport patches for CMake 4

emilazy committed:
graphite2: fix build with CMake 4

emilazy committed:
double-conversion: backport patches for CMake 4

emilazy committed:
woff2: fix build with CMake 4

emilazy committed:
libtiff: backport patches for CMake 4

emilazy committed:
soxr: fix build with CMake 4

emilazy committed:
libid3tag: backport patch for CMake 4

emilazy committed:
vid-stab: fix build with CMake 4

emilazy committed:
qhull: fix build with CMake 4

emilazy committed:
srt: backport patch for CMake 4

emilazy committed:
cjson: fix build with CMake 4

emilazy committed:
zopfli: fix build with CMake 4

emilazy committed:
libuchardet: backport patch for CMake 4

emilazy committed:
aften: fix build with CMake 4

emilazy committed:
x265: backport patches for CMake 4

emilazy committed:
gflags: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
cpptoml: fix build with CMake 4

emilazy committed:
cmake: drop sequence numbers from patches

These aren’t that helpful, since most of the patches don’t have
a strict ordering and a lot of them have been dropped anyway.

emilazy committed:
libipt: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
encfs: fix build with CMake 4

Deprecated upstream; won’t be fixed there.

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libspatialaudio: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
usrsctp: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
ldacbt: fix build with CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
xdg-utils-cxx: fix build with CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
taglib_1: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
ebook_tools: fix build with CMake 4

Not sure how to feel about Okular depending on this library that
hasn’t been touched since 2012 for EPUB support.

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
freeglut: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libebml: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
olm: fix build with CMake 4

Oh no, not this package again.

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libebur128: fix build with CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libmatroska: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
multimarkdown: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
uavs3d: fix build with CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
librdf_raptor2: fix build with CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
hotdoc: fix build with CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libappimage: backport patches for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libvncserver: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
kdePackages.sddm.unwrapped: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libcue: fix build with CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
libjxl: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy committed:
lzfse: fix build with CMake 4

emilazy committed:
blasfeo: backport patch for CMake 4

Co-authored-by: Guilhem Saurel <guilhem.saurel@laas.fr>

emilazy committed:
console-bridge: backport patch for CMake 4

Co-authored-by: Guilhem Saurel <guilhem.saurel@laas.fr>

emilazy committed:
qpoases: backport patch for CMake 4

Co-authored-by: Guilhem Saurel <guilhem.saurel@laas.fr>

emilazy committed:
scalapack: backport patch for CMake 4

Co-authored-by: Guilhem Saurel <guilhem.saurel@laas.fr>

emilazy committed:
crc32c: backport patch for CMake 4

Co-authored-by: Guilhem Saurel <guilhem.saurel@laas.fr>

emilazy committed:
lib3mf: backport patch for CMake 4

Co-authored-by: Guilhem Saurel <guilhem.saurel@laas.fr>

emilazy commented:
I’ve built a whole bunch of stuff on a local integration branch including this and https://github.com/NixOS/nixpkgs/pull/442167 now, and thanks to the efforts of @K900 and @nim65s we’ve also built the NixOS graphical ISO and more, and picked up a lot of fixes along the way. We’ll run through another quick round of builds now that stuff has been tidied up and integrated, but I think this will be ready to merge into `staging` just in time for the next cycle in the next day or two, at which point we’ll find all the *other* stuff that breaks and probably have some of the longest couple weeks on record.

Thanks @biodranik for sticking with this PR for so long, especially as a first contribution! I swear we usually get stuff merged faster than this; this one is an unusually wide‐impact breaking change and everyone interested in pushing this forward had a good few months juggling other priorities and, frankly, putting off dealing with the file‐mangling hook. I’m hopeful that with the amount of pre‐merge validation and fixing we’ve gone through in the past couple weeks this will go relatively smoothly through `staging-next`, but I still expect plenty of regressions in packages closer to the leaves of the tree that we’ll have to deal with during Zero Hydra Failures for the upcoming stable release.

I hope I am on a beach somewhere when they release CMake 5.

emilazy committed:
cfitsio: backport patch for CMake 4

Co-authored-by: K900 <me@0upti.me>

emilazy reviewed (approved):
It’s time!

trofi commented:
Bisect says 77b73f80712534cf6bb8d2c1c907e1157692f0d1 broke quite a few packages on` staging`. Some on my system that fail right now:

- `fuzzylite`: `Compatibility with CMake < 3.5 has been removed from CMake.`
- `libunarr`: `Compatibility with CMake < 3.5 has been removed from CMake.`
- `innoextract`: `Compatibility with CMake < 3.5 has been removed from CMake.`
- `flare`: `Compatibility with CMake < 3.5 has been removed from CMake.`
- `cdrkit`: ` No cmake_minimum_required command is present.  A line of code such as`
- `yajl`: `Compatibility with CMake < 3.5 has been removed from CMake.`
- `libvdpau-va-gl`: `Compatibility with CMake < 3.5 has been removed from CMake.`

K900 commented:
This is unfortunately unavoidable and we will have to just keep fixing those as they come up.

trofi commented:
In the past high-profile packages known to cause widespread failures occasionally had a separate jobset to weed out most failures. I think it avoided a bulk of failures and slightly improved bisectability of `nixpkgs`.

trofi commented:
> libvdpau-va-gl: Compatibility with CMake < 3.5 has been removed from CMake.

Proposed the fix as: https://github.com/NixOS/nixpkgs/pull/444579

trofi commented:
Proposed `fuzzylite` fix as:
- https://github.com/NixOS/nixpkgs/pull/449261
